radix-transactions/src/validation/transaction_validator.rs

```diff
@@ -379,12 +467,646 @@ impl TransactionValidator {
+                        SubintentValidationError::ChildSubintentNotIncludedInTransaction(child_hash)
+                    })?;
+                if subintent_details.parent == PLACEHOLDER_PARENT {
+                    subintent_details.parent = parent_hash;
+                } else {
+                    return Err(SubintentValidationError::SubintentHasMultipleParents(
+                        child_hash,
+                    ));
+                }
+                children_details.push(SubintentIndex(subintent_index));
+            }
+            all_subintent_details
+                .get_mut(&subintent_hash)
+                .unwrap()
+                .children = children_details;
+        }
+
+        // STEP 3
+        // ------
+        // We traverse the child relationships from the root, and mark a depth.
+        // We error if any exceed the maximum depth.
+        //
+        // As each child has at most one parent, we can guarantee the work is bounded
+        // by the total number of subintents.
+        let mut work_list = vec![];
+        for index in root_intent_details.children.iter() {
+            work_list.push((*index, 1));
+        }
+
+        loop {
+            let Some((index, depth)) = work_list.pop() else {
+                break;
+            };
+            if depth > self.config.max_subintent_depth {
+                let (hash, _) = all_subintent_details.get_index(index.0).unwrap();
+                return Err(SubintentValidationError::SubintentExceedsMaxDepth(*hash));
+            }
+            let (_, subintent_details) = all_subintent_details.get_index_mut(index.0).unwrap();
+            subintent_details.depth = depth;
+            for index in subintent_details.children.iter() {
+                work_list.push((*index, depth + 1));
+            }
+        }
+
+        // STEP 4
+        // ------
+        // We check that every subintent has a marked "depth from root".
+        //
+        // Combined with step 2 and step 3, we now have that:
+        // * Every subintent has a unique parent.
+        // * Every subintent is reachable from the root.
+        //
+        // Therefore there is a unique path from every subintent to the root
+        // So we have confirmed the subintents form a tree.
+        for (hash, details) in all_subintent_details.iter() {
+            if details.depth == 0 {
+                return Err(
+                    SubintentValidationError::SubintentIsNotReachableFromTheTransactionIntent(
+                        *hash,
+                    ),
+                );
+            }
+        }
+
+        Ok(IntentRelationships {
+            root_intent: root_intent_details,
+            subintents: all_subintent_details,
+        })
+    }
+}
+
+struct HeaderAggregation {
+    overall_start_epoch_inclusive: Epoch,
+    overall_end_epoch_exclusive: Epoch,
+    overall_start_timestamp_inclusive: Option<Instant>,
+    overall_end_timestamp_exclusive: Option<Instant>,
```

Comment:
Do we intend to validate the overall timestamp range?

Reply:
I believe this is validated in `HeaderAggregation::update`? Is there some other validation you've noticed that I've missed?